### ISSUE TITLE
multiple optimizer variablesI ( adam, adma_1) 

### ISSUE THREAD

**Issue by deepakmeena635** [deepmind/sonnet#118]
python3 
tf 1.9 
latest sonnet from pip3
ubuntu18
on cpu
 
please take a minute and a cookie to read the original comment [Here](https://github.com/tensorflow/tensorflow/issues/5786#issuecomment-461347077)  (includes all the relevent code) 

or below is an overview of the problem 

im creating sub graphs using sonnet and slim and  then putting them together in an Estimator's model function, the model seems to train without any errors/warning, but on running estimator_object.trainable_variables( ) i'm getting multiple betas and Adam variables, 

['beta1_power',
 'beta2_power',
 'embd/Conv/biases',
 'embd/Conv/biases/Adam',
 'embd/Conv/biases/Adam_1',
 'embd/Conv/weights',
 'embd/Conv/weights/Adam',
 'embd/Conv/weights/Adam_1',
 'global_step',
 'logit_mod/linear/b',
 'logit_mod/linear/b/Adam',
 'logit_mod/linear/b/Adam_1',
 'logit_mod/linear/w',
 'logit_mod/linear/w/Adam',
 'logit_mod/linear/w/Adam_1']

it seemed counter intutive, i did a google search and stumbled on the issue linked above, idk if its a problem just want to report if its an issue, i opened an issue here coz its possible that tf's developers arent famalier with internals of sonnet 

listing the questions/doubts here again:
**why are there two betas and "Adam" variables?, have i made any mistakes or this is because of using different modules(sonnet and slim) together ?**
any down sides of this?
my guess is the two modules will not be able to updaet gradients as supposed to and will learn independent of each other instead of one big composit function 

im not an expert of tf's low level  functionalities so any feedback will be appretiated, thanks here's something for u  :sweet_potato:  :)

**Comment by malcolmreynolds** [comment]
This is normal - for each trainable variable, Adam creates two auxiliary variables, for first and second order moments:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/adam.py#L131 - and similarly needs two additional scalars for the decay rates: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/adam.py#L124

You can see the same thing happening with the following toy example, to demonstrate that this is not something introduced by Sonnet/Slim mixing, or Estimator, or anything else.

```python
import sonnet as snt
import tensorflow as tf

input = tf.random_uniform([16, 784])
logits = snt.Linear(10)(input)
loss = tf.reduce_mean(logits)  # dummy loss function, still demonstrates the point
adam = tf.train.AdamOptimizer(learning_rate=1e-3)
train_op = adam.minimize(loss)

print(tf.get_collection(tf.GraphKeys.VARIABLES))
```

I get this output:

```
[<tf.Variable 'linear/w:0' shape=(784, 10) dtype=float32_ref>,
 <tf.Variable 'linear/b:0' shape=(10,) dtype=float32_ref>,
 <tf.Variable 'beta1_power:0' shape=() dtype=float32_ref>,
 <tf.Variable 'beta2_power:0' shape=() dtype=float32_ref>,
 <tf.Variable 'linear/w/Adam:0' shape=(784, 10) dtype=float32_ref>,
 <tf.Variable 'linear/w/Adam_1:0' shape=(784, 10) dtype=float32_ref>,
 <tf.Variable 'linear/b/Adam:0' shape=(10,) dtype=float32_ref>,
 <tf.Variable 'linear/b/Adam_1:0' shape=(10,) dtype=float32_ref>]
```

It's worth pointing out though that you should be somewhat careful when mixing different high level libraries - in particular, assumptions about how variables are shared could differ, so you are right to be checking whether the list of created variables looks correct. In this case though, I think you are fine.

Please reopen if you have any further questions!

**Comment by deepakmeena635** [comment]
i tested it the same day, with a simple example
thanks for explanation and for the advice
